tools/src/main/scala/scala/scalanative/linker/Link.scala

```diff
@@ -3,14 +3,33 @@ package linker
 
 import scalanative.nir._
 import scalanative.util.Scope
+import scala.scalanative.io.VirtualDirectory
+import java.nio.file.Path
 
 object Link {
 
+  private val cache =
+    collection.mutable.HashMap[(Path, Seq[Global], Seq[Long]), Result]()
+
   /** Load all clases and methods reachable from the entry points. */
   def apply(config: build.Config, entries: Seq[Global])(implicit
       scope: Scope
-  ): Result =
-    Reach(config, entries, ClassLoader.fromDisk(config))
+  ): Result = {
+    val classpath = config.classPath.map { path =>
+      ClassPath(VirtualDirectory.real(path))
+    }
+    val classPathMtime = classpath.map(_.lastModifiedMillis)
+    val key = (config.buildPath, entries, classPathMtime)
+    cache.get(key) match {
+      case None =>
+        val res = Reach(config, entries, ClassLoader.fromClasspath(classpath))
```

Comment:
I wonder if we could change cache granularity from the whole `linker.Result` for the whole build, to a classpath entry (single jar/directory). But that might require lots of other changes

Reply:
That was a change that was made to the backend where it goes by each classpath entry. We also started passing the artifacts from the `ll` phase to the `clang` backend as part of that change as well as the artifacts from each native lib to collect for the final link phase.